sockeye/lexicon.py

```diff
@@ -158,3 +183,108 @@ def _init_default(self, sym_name, arr):
         logger.info("Initializing '%s' with lexicon.", sym_name)
         assert len(arr.shape) == 2, "Only 2d weight matrices supported."
         self.lexicon.copyto(arr)
+
+
+class TopKLexicon:
+    """
+    Lexicon component that stores the k most likely target words for each source word.  Used during
+    decoding to restrict target vocabulary for each source sequence.
+
+    :param vocab_source: Trained model source vocabulary.
+    :param vocab_target: Trained mode target vocabulary.
+    """
+    def __init__(self,
+                 vocab_source: Dict[str, int],
+                 vocab_target: Dict[str, int]) -> None:
+        self.vocab_source = vocab_source
+        self.vocab_target = vocab_target
+        # Shape: (vocab_source_size, k), k determined at create() or load()
+        self.lex = None  # type: np.ndarray
+        # Always allow special vocab symbols in target vocab
+        self.always_allow = np.array([vocab_target[symbol] for symbol in C.VOCAB_SYMBOLS], dtype=np.int)
+
+    def create(self, path: str, k: int = 20):
+        """
+        Create from a scored lexicon file (fast_align format) using vocab from a trained Sockeye model.
+
+        :param path: Path to lexicon file.
+        :param k: Number of target entries per source to keep.
+        """
+        self.lex = np.zeros((len(self.vocab_source), k), dtype=np.int)
+        # Read lexicon
+        src_unk_id = self.vocab_source[C.UNK_SYMBOL]
+        trg_unk_id = self.vocab_target[C.UNK_SYMBOL]
+        _lex = collections.defaultdict(dict)  # type: Dict[int, Dict[int, float]]
+        for src_id, trg_id, prob in lexicon_iterator(path, self.vocab_source, self.vocab_target):
+            # Unk token will always be part of target vocab, so no need to track it here
+            if src_id == src_unk_id or trg_id == trg_unk_id:
+                continue
+            _lex[src_id][trg_id] = prob
+        # Sort and copy top-k trg_ids to lex array row src_id
+        for src_id, trg_entries in _lex.items():
+            top_k = list(sorted(trg_entries.items(), key=operator.itemgetter(1), reverse=True))[:k]
+            self.lex[src_id, :len(top_k)] = list(sorted(trg_id for (trg_id, _) in top_k))
+            # Free memory after copy
+            trg_entries.clear()
```

Comment:
would it make sense to clear _lex after this loop as well?

Reply:
We should be okay since _lex goes out of scope right after this.